linear_operator/operators/block_matrix_linear_operator.py

```diff
@@ -0,0 +1,168 @@
+from typing import List, Optional, Tuple, Union
+
+import torch
+from jaxtyping import Float
+from torch import Tensor
+
+from ._linear_operator import IndexType, LinearOperator
+from .dense_linear_operator import DenseLinearOperator
+from .zero_linear_operator import ZeroLinearOperator
+
+
+class BlockMatrixLinearOperator(LinearOperator):
+    """
+    A TxT block matrix of LinearOperators.
+
+    Idea. Represent [TN, TM] tensors by TxT blocks of NxM lazy tensors.
+
+    Implementation. A block linear operator class that can keep track of the [T, T] block structure,
+    represented as T^2 lazy tensors of the same shape. Implement matrix multiplication between block matrices as
+    the appropriate linear operators on the blocks.
+
+    :param linear_operators: A TxT nested list of linear operators representing a 2-D matrix
+    """
+
+    def __init__(self, linear_operators: List[List[LinearOperator]]) -> None:
+        assert isinstance(linear_operators, list), f"{self.__class__.__name__} expects a nested list of LinearOperators"
+        assert len(linear_operators) > 0, "must have non-empty list"
+        assert len(linear_operators[0]) == len(linear_operators), "must be square over block dimensions"
+
+        super().__init__(linear_operators)
+
+        self.linear_operators = linear_operators
+        self.num_tasks = len(self.linear_operators)
+        self.block_rows = linear_operators[0][0].shape[0]
+        self.block_cols = linear_operators[0][0].shape[1]
+
+        # Check that provided operators all have the same shape
+        T = self.num_tasks
+        for i in range(T):
+            for j in range(T):
+                assert (
+                    linear_operators[i][j].shape[0] == self.block_rows
+                ), "the number of rows much match for all linear operators"
+                assert (
+                    linear_operators[i][j].shape[1] == self.block_cols
+                ), "the number of columns much match for all linear operators"
```

Comment:
Other small nit: this is a bit of checking that's happening under the hood, which might get a bit expensive. Would it be possible to do this checking only if we're in debug mode (using the context manager `with linear_operator.settings.debug(True)`?

If that's undesirable I'm happy to keep it as is, but just seeing if there's ways to lower the python overhead.

Reply:
Makes sense. Thanks for the suggestion!